chain/consensus/mir/manager.go

```diff
@@ -54,108 +69,178 @@ func NewManager(ctx context.Context, addr address.Address, api v1api.FullNode) (
 	}
+
+	memberships := make([]map[t.NodeID]t.NodeAddress, ConfigOffset+1)
+	for i := 0; i < ConfigOffset+1; i++ {
+		memberships[t.EpochNr(i)] = initialMembership
+	}
+
 	mirID := newMirID(subnetID.String(), addr.String())
-	mirAddr := nodeAddrs[t.NodeID(mirID)]
+	mirAddr := initialMembership[t.NodeID(mirID)]
 
 	log.Info("Eudico node's Mir ID: ", mirID)
 	log.Info("Eudico node's address in Mir: ", mirAddr)
 	log.Info("Mir nodes IDs: ", nodeIDs)
-	log.Info("Mir nodes addresses: ", nodeAddrs)
+	log.Info("Mir nodes addresses: ", initialMembership)
 
 	peerID, err := peer.AddrInfoFromP2pAddr(mirAddr)
 	if err != nil {
 		return nil, fmt.Errorf("failed to get addr info: %w", err)
 	}
 
 	h, err := libp2p.New(
-		libp2p.Identity(hostPrivKey),
+		libp2p.Identity(libp2pKey),
 		libp2p.DefaultTransports,
 		libp2p.ListenAddrs(peerID.Addrs[0]),
 	)
 	if err != nil {
-		return nil, fmt.Errorf("failed to construct libp2p host: %w", err)
+		return nil, fmt.Errorf("failed to create libp2p host: %w", err)
 	}
 
 	// Create Mir modules.
 
-	netTransport := mirlibp2p.NewTransport(h, nodeAddrs, t.NodeID(mirID), newMirLogger(managerLog))
-	if err := netTransport.Start(); err != nil {
-		return nil, fmt.Errorf("failed to create libp2p transport: %w", err)
-	}
-	netTransport.Connect(ctx)
-
 	wal, err := NewWAL(mirID, fmt.Sprintf("eudico-wal-%s", addr))
 	if err != nil {
 		return nil, fmt.Errorf("failed to create WAL: %w", err)
 	}
 
-	// Instantiate the ISS protocol module with default configuration.
-	issConfig := iss.DefaultConfig(nodeIDs)
-	issProtocol, err := iss.New(t.NodeID(mirID), issConfig, newMirLogger(managerLog))
+	netTransport, err := mirlibp2p.NewTransport(h, t.NodeID(mirID), newMirLogger(managerLog))
 	if err != nil {
-		return nil, fmt.Errorf("could not instantiate ISS protocol module: %w", err)
+		return nil, fmt.Errorf("failed to create transport: %w", err)
+	}
+	if err := netTransport.Start(); err != nil {
+		return nil, fmt.Errorf("failed to start transport: %w", err)
 	}
+	netTransport.Connect(ctx, initialMembership)
 
 	cryptoManager, err := NewCryptoManager(addr, api)
 	if err != nil {
 		return nil, fmt.Errorf("failed to create crypto manager: %w", err)
 	}
 
-	sm := NewStateManager()
+	// Instantiate the ISS protocol module with default configuration.
 
-	pool := newRequestPool()
+	issConfig := iss.DefaultConfig(initialMembership)
+	issConfig.ConfigOffset = ConfigOffset
+	issProtocol, err := iss.New(
+		t.NodeID(mirID),
+		iss.DefaultModuleConfig(),
+		issConfig,
+		iss.InitialStateSnapshot([]byte{}, issConfig),
+		newMirLogger(managerLog),
+	)
+	if err != nil {
+		return nil, fmt.Errorf("could not instantiate ISS protocol module: %w", err)
+	}
+
+	// Use a simple mempool for incoming requests.
+	mempool := simplemempool.NewModule(
+		&simplemempool.ModuleConfig{
+			Self:   "mempool",
+			Hasher: "hasher",
+		},
+		&simplemempool.ModuleParams{
+			MaxTransactionsInBatch: 10,
+		},
+	)
+
+	mscFactory := factorymodule.New(
```

Comment:
Ok, I didn't get a thing of this, I may need to learn once and for all the Mir base code. If @matejpavlovic doesn't see anything weird here, LGTM :)

Reply:
Actually, that's @matejpavlovic 's reference code for using the availability layer.

Reply:
@matejpavlovic, can I invite you for a coffee/beer in exchange for a quick overview of the Mir base code? I feel I can be significantly more helpful and less annoying in my reviews if I steal an hour of your time for this. Thanks.

Reply:
The ugliness of the code (to a large extent) fixed by https://github.com/filecoin-project/mir/pull/205

Reply:
Amazing!!! Thank you, @matejpavlovic.